contrib/mesos/pkg/node/registrator.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"reflect"
+	"time"
+
+	log "github.com/golang/glog"
+	"k8s.io/kubernetes/contrib/mesos/pkg/queue"
+	"k8s.io/kubernetes/contrib/mesos/pkg/runtime"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+)
+
+const (
+	defaultInboxSize = 1000
+)
+
+type Registrator interface {
+	Register(hostName string, labels map[string]string) (bool, error)
+	Run(terminate <-chan struct{}) error
+}
+
+type registration struct {
+	hostName string
+	labels   map[string]string
+}
+
+func (r *registration) Copy() queue.Copyable {
+	return &registration{
+		hostName: r.hostName,
+		labels:   r.labels, // labels are never changed, hence just copy the slice
+	}
+}
+
+func (r *registration) GetUID() string {
+	return r.hostName
+}
+
+func (r *registration) Value() queue.UniqueCopyable {
+	return r
+}
+
+func (r *registration) Is(types queue.EventType) bool {
```

Comment:
inbox is misused here. the chan passed to the HistoricalFIFO is to be
written to by the fifo and read from by a client (that wants each and every
event, not just the merged version that's popped). this implementation
(registrator) doesn't need those details, it actually wants the merged copy
of things so the chan isn't necessary. and the client (e.g. registrator)
definitely should not be writing to the chan under any circumstances.

On Thu, Sep 17, 2015 at 10:55 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/node/registrator.go
> <https://github.com/kubernetes/kubernetes/pull/13857#discussion_r39754844>
> :
>
> > +func (r *registration) Copy() queue.Copyable {
> > +	return &registration{
> > +		hostName: r.hostName,
> > +		labels:   r.labels, // labels are never changed, hence just copy the slice
> > +	}
> > +}
> > +
> > +func (r *registration) GetUID() string {
> > +	return r.hostName
> > +}
> > +
> > +func (r *registration) Value() queue.UniqueCopyable {
> > +	return r
> > +}
> > +
> > +func (r *registration) Is(types queue.EventType) bool {
>
> Without the inbox yes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13857/files#r39754844>.
>